llpc/context/llpcCompiler.cpp

```diff
@@ -1233,31 +1237,59 @@ Result Compiler::buildPipelineInternal(Context *context, ArrayRef<const Pipeline
       if (!shaderInfoEntry || !shaderInfoEntry->pModuleData || (stageSkipMask & shaderStageToMask(entryStage)))
         continue;
 
-      std::unique_ptr<lgc::LegacyPassManager> lowerPassMgr(lgc::LegacyPassManager::Create());
-      lowerPassMgr->setPassIndex(&passIndex);
+      if (cl::LlvmNewPM) {
+        std::unique_ptr<lgc::PassManager> lowerPassMgr(lgc::PassManager::Create());
+        lowerPassMgr->setPassIndex(&passIndex);
 
-      // Set the shader stage in the Builder.
-      context->getBuilder()->setShaderStage(getLgcShaderStage(entryStage));
+        // Set the shader stage in the Builder.
+        context->getBuilder()->setShaderStage(getLgcShaderStage(entryStage));
 
-      // Start timer for translate.
-      timerProfiler.addTimerStartStopPass(&*lowerPassMgr, TimerTranslate, true);
+        // Start timer for translate.
+        timerProfiler.addTimerStartStopPass(*lowerPassMgr, TimerTranslate, true);
 
-      // SPIR-V translation, then dump the result.
-      lowerPassMgr->add(createSpirvLowerTranslator(entryStage, shaderInfoEntry));
-      if (EnableOuts()) {
-        lowerPassMgr->add(createPrintModulePass(
-            outs(), "\n"
-                    "===============================================================================\n"
-                    "// LLPC SPIRV-to-LLVM translation results\n"));
-      }
-      // Stop timer for translate.
-      timerProfiler.addTimerStartStopPass(&*lowerPassMgr, TimerTranslate, false);
+        // SPIR-V translation, then dump the result.
+        lowerPassMgr->addPass(SpirvLowerTranslator(entryStage, shaderInfoEntry));
+        if (EnableOuts()) {
+          lowerPassMgr->addPass(PrintModulePass(
+              outs(), "\n"
+                      "===============================================================================\n"
+                      "// LLPC SPIRV-to-LLVM translation results\n"));
+        }
+        // Stop timer for translate.
+        timerProfiler.addTimerStartStopPass(*lowerPassMgr, TimerTranslate, false);
 
-      // Run the passes.
-      bool success = runPasses(&*lowerPassMgr, modules[shaderIndex]);
-      if (!success) {
-        LLPC_ERRS("Failed to translate SPIR-V or run per-shader passes\n");
-        result = Result::ErrorInvalidShader;
+        bool success = runPasses(&*lowerPassMgr, modules[shaderIndex]);
+        if (!success) {
+          LLPC_ERRS("Failed to translate SPIR-V or run per-shader passes\n");
+          result = Result::ErrorInvalidShader;
+        }
+      } else {
+        std::unique_ptr<lgc::LegacyPassManager> lowerPassMgr(lgc::LegacyPassManager::Create());
```

Comment:
It is unfortunate that we need to duplicate quite a lot of code, but the way of adding the passes is so different that it makes sense to do that.